File: testing_codes/Python/1. FTP.py
import socket
import os
import threading

class FileServer:
    def __init__(self, port, file_directory):
        self.port = port
        self.file_directory = file_directory
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_socket.bind(('localhost', port))
        self.server_socket.listen(5)

    def start(self):
        while True:
            client_socket, _ = self.server_socket.accept()
            threading.Thread(target=self.handle_client, args=(client_socket,)).start()

    def handle_client(self, client_socket):
        with client_socket:
            data = client_socket.recv(1024).decode().strip()
            if data:
                filename = os.path.join(self.file_directory, data)
                if os.path.exists(filename) and os.path.isfile(filename):
                    with open(filename, 'rb') as file:
                        data = file.read(1024)
                        while data:
                            client_socket.send(data)
                            data = file.read(1024)
                else:
                    client_socket.sendall(b"HTTP/1.1 404 Not Found\r\nConnection: close\r\n\r\n")


def main():
    import sys
    if len(sys.argv) != 3:
        print("Usage: python file_server.py <port> <file_directory>")
        return

    port = int(sys.argv[1])
    file_directory = sys.argv[2]

    server = FileServer(port, file_directory)
    server.start()


if __name__ == "__main__":
    main()

File: testing_codes/Java/1. FTP.java
import java.io.*;
import java.net.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class FileServer {

    private static final int THREAD_POOL_SIZE = 10;

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("Usage: java FileServer <port> <file_directory>");
            return;
        }

        int port = Integer.parseInt(args[0]);
        String fileDirectory = args[1];

        ExecutorService executorService = Executors.newFixedThreadPool(THREAD_POOL_SIZE);

        try (ServerSocket serverSocket = new ServerSocket(port)) {
            while (true) {
                Socket clientSocket = serverSocket.accept();
                executorService.submit(new Session(clientSocket, fileDirectory));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

class Session implements Runnable {

    private Socket socket;
    private String fileDirectory;

    public Session(Socket socket, String fileDirectory) {
        this.socket = socket;
        this.fileDirectory = fileDirectory;
    }

    @Override
    public void run() {
        try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {

            String filename = in.readLine();
            if (filename != null) {
                filename = filename.trim();
                File file = new File(fileDirectory, filename);
                if (file.exists() && file.isFile()) {
                    try (BufferedInputStream fileInputStream = new BufferedInputStream(new FileInputStream(file))) {
                        byte[] buffer = new byte[4096];
                        int bytesRead;
                        while ((bytesRead = fileInputStream.read(buffer)) != -1) {
                            socket.getOutputStream().write(buffer, 0, bytesRead);
                        }
                    }
                } else {
                    out.println("HTTP/1.1 404 Not Found\r\nConnection: close\r\n\r\n");
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                socket.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}
